system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,468 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "SCNetworkService{{ id: {:?}, name: {:?}, enabled: {}, interface: {:?}, dns: {:?} }}",
+            self.id(),
+            self.name(),
+            self.enabled(),
+            self.interface(),
+            self.dns()
+        )
+    }
+}
+
+
+declare_TCFType!{
+    /// Network interface object.
+    SCNetworkInterface, SCNetworkInterfaceRef
+}
+
+impl_TCFType!(
+    SCNetworkInterface,
+    SCNetworkInterfaceRef,
+    SCNetworkInterfaceGetTypeID
+);
+
+
+impl SCNetworkInterface {
+    /// Returns primary network interface
+    pub fn global(store: &SCDynamicStore) -> Option<Self> {
+        if let Some(ifname) = global_query(store, "PrimaryInterface") {
+            for iface in SCNetworkInterface::list() {
+                if let Some(bsd_name) = iface.bsd_name() {
+                    if bsd_name == ifname {
+                        return Some(iface);
+                    }
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns all network-capable interfaces on the system.
+    pub fn list() -> Vec<SCNetworkInterface> {
+        let array: CFArray<SCNetworkInterfaceRef> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkInterfaceCopyAll()) };
+
+        array
+            .get_all_values()
+            .iter()
+            .map(|interface_ptr| unsafe {
+                SCNetworkInterface::wrap_under_get_rule(*interface_ptr as _)
+            })
+            .collect::<Vec<SCNetworkInterface>>()
+    }
+
+    /// Returns the current MTU setting and the range of allowable values
+    pub fn mtu(&self) -> Option<SCNetworkInterfaceMtu> {
+        let mut current = 0i32;
+        let mut min = 0i32;
+        let mut max = 0i32;
+
+        let ret_code =
+            unsafe { SCNetworkInterfaceCopyMTU(self.0, &mut current, &mut min, &mut max) };
+        if ret_code == 0 {
+            None
+        } else {
+            Some(SCNetworkInterfaceMtu {
+                current: current as u32,
+                min: if min < 0 { None } else { Some(min as u32) },
+                max: if max < 0 { None } else { Some(max as u32) },
+            })
+        }
+    }
+
+    /// Returns the BSD interface or device name
+    pub fn bsd_name(&self) -> Option<String> {
+        unsafe {
+            let str_ptr = SCNetworkInterfaceGetBSDName(self.0);
+            if str_ptr.is_null() {
+                None
+            } else {
+                Some(CFString::wrap_under_get_rule(str_ptr).to_string())
+            }
+        }
+    }
+
+    /// Returns the BSD interface or device name
+    pub fn name(&self) -> Option<String> {
+        self.bsd_name()
+    }
+
+    /// Returns the network interface type
+    pub fn type_(&self) -> Option<String> {
+        unsafe {
+            let str_ptr = SCNetworkInterfaceGetInterfaceType(self.0);
+            if str_ptr.is_null() {
+                None
```

Comment:
The `SCNetworkInterfaceGetInterfaceType` documentation does not mention it can return `NULL`. So it's probably safe to assume it doesn't and change the return type here to `-> String` directly.